plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,412 @@
+    fu_udev_device_pwrite_full (FU_UDEV_DEVICE (device), 0, buf, sizeof(struct cmd_fw_upgrade) + 2, error);
+
+    /* read fw updgrad command result */
+    g_clear_error (error);
+    memset(buf, 0,PXI_OTA_BUF_SZ);
+    buf[0] = PXI_HID_DEV_OTA_INPUT_REPORT_ID;
+
+    fu_udev_device_pread_full (FU_UDEV_DEVICE (device), 1, buf, PXI_OTA_BUF_SZ,error);
+    fu_common_dump_raw (G_LOG_DOMAIN, "buf", (guint8 *)buf, PXI_OTA_BUF_SZ);
+    fu_memcpy_safe((guint8 *)&fw_notify_result, sizeof(struct ret_fw_notify), 0, buf,PXI_OTA_BUF_SZ,1, sizeof(struct ret_fw_notify), error);
+
+    if(fw_notify_result.opcode != CMD_FW_UPGRADE)
+    {
+        g_set_error (error, FWUPD_ERROR,FWUPD_ERROR_READ, "FWUPD_ERROR_READ %d", CMD_FW_WRITE);
+        return FALSE;
+    }
+
+    fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+    /* send device reset command */
+    g_clear_error (error);
+    memset(buf, 0,PXI_OTA_BUF_SZ);
+    buf[0] = PXI_HID_DEV_OTA_OUTPUT_REPORT_ID;
+    buf[1] = CMD_FW_MCU_RESET;
+
+    fw_ota_disconnect.reason = g_bytes_get_size(fw);
+    fu_memcpy_safe(buf, PXI_OTA_BUF_SZ, 2, (guint8*)&fw_ota_disconnect, sizeof(struct cmd_fw_ota_disconnect), 0, sizeof(struct cmd_fw_ota_disconnect), error);
+    fu_udev_device_pwrite_full (FU_UDEV_DEVICE (device), 0, buf, sizeof(struct cmd_fw_ota_disconnect) + 2, error);
+
+
+    return TRUE;
+
+
+}
+
+
+static gboolean
+fu_pxi_device_probe (FuDevice *device, GError **error)
+{
+    /* set the physical ID */
+    if (!fu_udev_device_set_physical_id (FU_UDEV_DEVICE (device), "hid", error))
+        return FALSE;
+
+    return TRUE;
+}
+
+static gboolean
+fu_pxi_device_setup (FuDevice *device, GError **error)
+{
+
+
+    guint8 buf[64];
+
+    g_autofree gchar *current_version = NULL;
+
+    struct hci_evt evt;
+
+
+    buf[0] = PXI_HID_DEV_OTA_OUTPUT_REPORT_ID;
+    buf[1] = CMD_FW_OTA_INIT;
+    fu_udev_device_pwrite_full (FU_UDEV_DEVICE (device), 0, (guint8 *)buf, 2, error);
+
+
+    buf[0] = PXI_HID_DEV_OTA_OUTPUT_REPORT_ID;
+    buf[1] = CMD_FW_GET_INFO;
+    fu_udev_device_pwrite_full (FU_UDEV_DEVICE (device), 0, (guint8 *)buf, 2, error);
+
+
+    buf[0] = PXI_HID_DEV_OTA_FEATURE_REPORT_ID;
+    buf[1] = CMD_FW_GET_INFO;
+    fu_pxi_device_get_feature (device, buf, 64, error);
+
+    fu_common_dump_raw (G_LOG_DOMAIN, "buf", (guint8 *)buf, 64);
+
+    fu_memcpy_safe((guint8 *)&evt, sizeof(struct hci_evt), 0, buf, 64, 0, sizeof(struct hci_evt), error);
+
+    if(evt.evt_param.opcode != CMD_FW_GET_INFO)
+        return FALSE;
+
+
+    current_version = g_strndup ((gchar*)&evt.evt_param.ret_param.fw_info_get.version, 5);
+
+    g_debug("checksum %x",evt.evt_param.ret_param.fw_info_get.checksum);
+    g_debug("version %s",(gchar*)evt.evt_param.ret_param.fw_info_get.version);
+    g_debug("current_version %s",current_version);
+    fu_device_set_version (device, current_version);
+
+    fu_device_add_checksum (device, g_strdup_printf("0x%4x",evt.evt_param.ret_param.fw_info_get.checksum));
```

Comment:
this leaks the g_strdup_printf -- you need to assign it to `g_autofree gchar *tmp` or something.